Block-builder: alerts and dashboard updates (#12118)
grafana/mimir: docs/sources/mimir/manage/mimir-runbooks/_index.md

- If the `block-builder-scheduler.lookback-on-no-commit` does not cover the time when the issue started, set it long enough so that these new block-builders start back far enough to cover the missing data.
- Investigate why the block-builder fails, while the ingesters, who consumed the same data, don't.

#### MimirBlockBuilderDataSkipped

This alert fires when the block-builder-scheduler has detected a gap in either committed jobs or planned jobs.

How it **works**:

> - Block-builder-scheduler is in charge of both planning "jobs" for block-builders to consume, as well as advancing the per-partition commit when one of these jobs is completed. Each job has a start offset (inclusive) and an end offset (exclusive).

Reviewer: ```suggestion
- Block-builder-scheduler is in charge of both planning "jobs" for block-builders to consume, as well as advancing the per-partition commit when one of these jobs is completed. Each job has a start offset, which is inclusive, and an end offset, which is exclusive.
```

> - Block-builder-scheduler also verifies that neither planning nor committing jobs ever produces a gap. This alert fires when the scheduler detects one of these gaps.

Reviewer: ```suggestion
- Block-builder-scheduler also verifies that neither planning nor committing jobs ever produce a gap. This alert fires when the scheduler detects one of these gaps.
```

> - Note this is a "should never happen" problem, and will only happen if there's a correctness bug.

Reviewer: ```suggestion
- Note that this problem only happens if there's a correctness bug.
```

Reviewer: I'm not totally clear on the meaning of "correctness bug". Is this a term Mimir users would be familiar with?

Author: There are bugs that crash, bugs that cause high resource usage, [skip many other classes of bugs...] and then bugs that cause an algorithm to work incorrectly. This would be the latter.

Good point- here we can just call it a bug in the planning logic.

How to **investigate**:

> - Look at the block-builder-scheduler logs around the affected time. There will be a log line that mentions "gap detected in offset advancement" with relevant data.

Reviewer: ```suggestion
- Look at the block-builder-scheduler logs around the affected time. There's a log line that mentions "gap detected in offset advancement" with relevant data.
```

> - Using this data, further look at block-builder-scheduler logs to find the jobs either planned or committed around this time, and see if you can piece together how the gap occurred.

Reviewer: ```suggestion
- Using this data, further look at block-builder-scheduler logs to find the jobs either planned or committed around this time, and see if you can piece together how the gap occurred.
```
```suggestion
- Using this data, continue to look at block-builder-scheduler logs to find the jobs either planned or committed around this time, and see if you can determine how the gap occurred.
```

Data recovery / temporary mitigation:

> You'll need to coax block-builder into consuming the skipped data. See the section under "Data recovery" for the `MimirBlockBuilderHasNotShippedBlocks` alert.

Reviewer: ```suggestion
You need to coax block-builder into consuming the skipped data. Refer to the section under "Data recovery" for the `MimirBlockBuilderHasNotShippedBlocks` alert.
```

Reviewer: "Coax" appears to personify the block-builder. Is there a difference word we could use? Perhaps simply `You need to make block-builder consume the skipped data.`
